### ISSUE TITLE
Character Highlighting in PDF

### ISSUE THREAD

**Issue by segreeeen** [piersdeseilligny/betterfountain#65]
It would be nice, if it was possible to generate PDFs for a specific Character (or maybe multiple Characters?) which is then highlighted in the PDF. This would make it even easier for the actors to find their lines. Following up this thought, in the future, this leads to a batch pdf-creation of individual PDFs for every character. In every individual pdf, the according character is highlighted. 

Of course it should be possible to enable/disable this option. There would have to be some configuration popup in which you have to choose the characters for which you'd like a personalized PDF in a multiple choice list, including a "select all" button, to batch-create the individual actor-PDFs or whatever this would be called.

Just some input, if you are up to the task - I know a lot of people who'd be happy to use that :). Carry on with the great work.

**Comment by daryluren** [comment]
I like this idea, and wanna point out, even though it makes things harder:

That for animations/audio dramas in particular, a single actor could play multiple characters. They might want their personalised PDF to only contain scenes/sequences that have any of their characters involved.

**Comment by segreeeen** [comment]
@daryluren yes, absolutely, that's why I suggested the multiple-choice list if characters. So you can choose the characters that should be highlighted in the pdf :). It would make sense to highlight different characters in different colours though, right?

**Comment by piersdeseilligny** [comment]
This all sounds really good - the main difficulty with implementing this doesn't seem to be technical though, it's more about planning the best way to configure this potential feature in a simple and elegant way. The two options I can imagine right now are:


* Implementing it as a new title page key, like the "Font:" one - the fountain document would include `Highlight characters: Alice, Bob` alongside the other title page keys. This makes the fact that a document is being highlighted extremely clear and simple, at the cost of less configurability (such as batch creation of PDFs, which couldn't really be implemented this way) and extending the `.fountain` spec in a non-standard way.

* Have a separate pdf configuration file, maybe with an extension like `.fountainexport`, alongside the `.fountain` one. This file could be editable with a [custom editor](https://code.visualstudio.com/api/extension-guides/custom-editors) and would contain a bunch of settings such as which characters are highlighted, in which color, if PDFs should be batch-created etc... the "Export PDF" command would then simply open the `.fountainexport` file (creating it first if it doesn't exist). This would allow for some pretty advanced and powerful features (such as #14), at the cost of simplicity.

What are your thoughts? And can anyone think of a better way of doing this, because neither really seems ideal :/

**Comment by segreeeen** [comment]
Hi @piersdeseilligny - I'd suggest not to overcomplicate this in trying to solve this in a generic matter, since the feature - as I conceived it, at least - would be specific to the VS Code Extension - i.e. not an addition to the fountain standard, but a BetterFountain feature. In my oppinion it would suffice to have a popup before exporting with a List of Characters which are chosable in a multiselect manner. This way there is no additional configuration necessary and very easy to define what you want. It also would not make too much sense, I think, to have title page keys or new config files. As I said, the feature is BetterFountain specific, and I doubt other editors would adopt either of these (i.e. title key or .fountainexport).

Made a little image to clarify what I meant: 
![image](https://user-images.githubusercontent.com/4742136/85223184-16c09000-b3c1-11ea-80aa-38a3372afd5a.png)

Just as a rough idea what I meant, this would look different in the end of course. You can choose which characters to export, or just select all (eventually also deselect all). I'm not saying that's the best way to do it, just probably way less drama than implementing a new fountain syntax feature or something. That would only make sense if other editors would also adopt the ways it's done. Let's keep it simple, maybe.

**Comment by daryluren** [comment]
At least if it were a config file it could be as complicated as the user would like, for instance if there were multiple actors who wanted personalised pdfs you could re-print them all every time you updated the script.
(pseudo json, just spitballing)
```
"pdf" : [ {
    /* standard one */
},{
    "filename_extension" : "bar",
    "filter_roles" : [ {"Alice", "#ff8040" }, "Bob"]
},{
    "filename_extension" : "foo",
    "filter_roles" : ["Jane", "John"]
}]
```

**Comment by daryluren** [comment]
Thinking about it, there's definitely still some technical things to worry about.

Some scripts might quickly switch between different scenes in a way that any role filtering demanded would be at the act/sequence level. And according to the fountain spec, the ##sections can be as ###deep as you ####want.

Another problem is detecting whether a character is in a Scene if they haven't got any dialogue. Is there a way to do that yet? The script convention seems to be to introduce characters in Action in all upper case. Would it be too audacious to demand betterfountain users to provide character names in upper case (even if they're already introduced) to ensure they're detected in these cases? This would help #14 too.

**Comment by segreeeen** [comment]
Right, totally overlooked that. Maybe such a structure could be kept in the background, and have the easy interface as well? I doubt that the majority of the users understands json files? I can be wrong though... In that case the .fountainexport file could be the nicer way of doing this? It's way more effort though.

**Comment by segreeeen** [comment]
@daryluren okay maybe we need some clarification first. My original thoght was just to highlight the characters names as in Fountain Syntax Characters, you know:

ALBERT
Dialoge of albert.

Just the ALBERT would be highlighted, not the "albert" in his dialogue text. Maybe we have different concepts in mind?

**Comment by piersdeseilligny** [comment]
> Right, totally overlooked that. Maybe such a structure could be kept in the background, and have the easy interface as well? I doubt that the majority of the users understands json files? I can be wrong though... In that case the .fountainexport file could be the nicer way of doing this? It's way more effort though.

Yeah I definitely agree that there should be no need to directly edit json files, but that's where a custom editor for a .fountainexport file would come into play. `.fountainexport` would literally just be a json file, but opening it in vscode would open an editor very similar to the "PDF Export" popup you suggested (as unfortunately, I don't think vscode allows extensions to contribute popups)

**Comment by piersdeseilligny** [comment]
> Thinking about it, there's definitely still some technical things to worry about.
> 
> Some scripts might quickly switch between different scenes in a way that any role filtering demanded would be at the act/sequence level. And according to the fountain spec, the ##sections can be as ###deep as you ####want.
> 
> Another problem is detecting whether a character is in a Scene if they haven't got any dialogue. Is there a way to do that yet? The script convention seems to be to introduce characters in Action in all upper case. Would it be too audacious to demand betterfountain users to provide character names in upper case (even if they're already introduced) to ensure they're detected in these cases? This would help #14 too.

Why would it be necessary to detect if a character is in scenes/sections though, wouldn't this feature simply highlight lines of dialogue? Or are you talking about completely filtering out scenes where a certain character isn't there?

**Comment by daryluren** [comment]
@segreeeen @piersdeseilligny 
Hi, yes sorry, you're right. "personalisation" had me imagining cocky actors who only want the pages that involve them. Lost the plot lol.

**Comment by piersdeseilligny** [comment]
haha yeah let's not make it even easier for actors to be lazy and avoid reading the entire screenplay. Highlighting character lines could be particularly useful for stuff like table reads though.

The way i'm suggesting we do this is:

* Start by implementing a command-based system, where one could run a command called something like "Export Screenplay PDF (highlighted character)", which would use `vscode.window.showQuickPick` to allow the user to select one or more characters to highlight, and then ask for which color to highlight them in. Even if we later end up implementing `.fountainexport` this command should probably exist anyway, and it'll probably be fairly simple to add.

* Carefully plan out a `fountainexport` spec that will allow any combination of printing profiles (that can override default settings) and additional options, such as character highlighting. I've started doing this [here](https://github.com/piersdeseilligny/betterfountain/blob/master/Fountainexport%20spec%20draft.md) (and by start, i mean written one small paragraph).

* Once we know exactly what fountainexport should do and how, we can write a JSON schema which outlines it and implement the "interpreter" within betterfountain. This can then be released as part of an update.

* Finally, we can add a custom editor for fountainexport files, which will allow editing these advanced export settings in an actually intuitive and pleasant way, rather than by editing JSON.

Please do let me know if you have any thoughts on this :) I'm currently busy working on improvements to the statistics page, which will eventually allow the creation of screenplay reports (as suggested in #14) much like Final Draft allows. Small sneak peak here:

<img width="200" src="https://user-images.githubusercontent.com/42496508/85320377-42c43a00-b4c3-11ea-89a1-5119d8008a7b.png"></img>

In light of these PDF Reports, adding a .fountainexport seems like it could actually be pretty useful

**Comment by daryluren** [comment]
I'm glad it's turning into an enhanced printing features module. If would be overkill if just for the colour.

**Comment by segreeeen** [comment]
Hey, sorry I'm so late to answer again. @piersdeseilligny I like the ideas you mentioned, if you like we could meet in discord voicechat to better plan the specifics. If you want, I think that would be easier than writing here. Maybe you could create a discord server for betterfountain? :)

**Comment by piersdeseilligny** [comment]
@segreeeen that's a good idea! I'm currently somewhere with an absolutely horrendous internet connection and i'm pretty busy these days, so i'm not sure about a voicechat, but I've made a discord server anyway: https://discord.gg/QBm7tKQ

**Comment by daryluren** [comment]
I was thinking if someone were to print the pdf, they may not want to print in colour for $$ reasons. Maybe there could be an option for the character highlight to be represented by modifying the border, say a double line border on the left side.

**Comment by piersdeseilligny** [comment]
Fixed by 1.7.0 thanks to @rnwzd 's amazing PR!
